tico/utils/pytree_utils.py

```diff
@@ -0,0 +1,134 @@
+import threading
+
+import torch
+from packaging.version import Version
+
+from tico.utils import logging
+from tico.utils.installed_packages import is_transformers_installed
+
+__all__ = ["register_dynamic_cache"]
+
+
+def register_dynamic_cache():
+    PyTreeRegistryHelper().register_dynamic_cache()
+
+
+class PyTreeRegistryHelper:
+    """
+    Thread-safe singleton helper class for registering custom PyTree nodes.
+
+    This class provides functionality to register DynamicCache as a PyTree node
+    for torch.export compatibility. This registration is only needed for
+    transformers versions below 4.50.0.
+
+    Thread Safety:
+    - Uses a class-level threading.Lock() to ensure thread-safe singleton instantiation
+    - Uses the same lock to protect the registration process from concurrent calls
+    """
+
+    _instance = None  # Class variable to hold the singleton instance
+    _has_called = False  # Flag to track if registration has been performed
+    _lock = threading.Lock()  # Class-level lock for thread-safe operations
+
+    def __init__(self):
+        """Private constructor to prevent direct instantiation"""
+        pass
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Thread-safe singleton instance creation using double-checked locking pattern.
+
+        Returns:
+            PyTreeRegistryHelper: The singleton instance of this class
+        """
+        if not cls._instance:
+            with cls._lock:  # Acquire lock for thread-safe instantiation
+                if not cls._instance:  # Double-check after acquiring lock
+                    cls._instance = super().__new__(cls)
+        return cls._instance
+
+    def register_dynamic_cache(self):
+        """
+        Registers DynamicCache as a PyTree node for torch.export compatibility.
+
+        This method is thread-safe and idempotent - it will only perform the
+        registration once, even if called multiple times from different threads.
+
+        Note:
+            This registration is only needed for transformers versions below 4.50.0.
+
+        Raises:
+            ImportError: If transformers package is not installed
+        """
+        with self._lock:  # Acquire lock for thread-safe registration
+            if self.__class__._has_called:
+                logger = logging.getLogger(__name__)
+                logger.debug("register_dynamic_cache already called, skipping")
+                return
+
+            self.__class__._has_called = True
+            logger = logging.getLogger(__name__)
+            logger.info("Registering DynamicCache PyTree node")
+
+        if not is_transformers_installed:  # type: ignore[truthy-function]
+            raise ImportError("transformers package is not installed")
+
+        import transformers
```

Comment:
`tico` does not assume transformers as dependency package, this guard is required.